File: KiCAD-MCP-Server/src/tools/symbol-creator.ts
/**
 * Symbol creator tools for KiCAD MCP server
 *
 * create_symbol              – add a new symbol to a .kicad_sym library
 * delete_symbol              – remove a symbol from a library
 * list_symbols_in_library    – list all symbols in a .kicad_sym file
 * register_symbol_library    – add library to sym-lib-table
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

const PinSchema = z.object({
  name: z.string().describe("Pin name, e.g. 'VCC', 'GND', 'IN+', '~' for unnamed"),
  number: z.union([z.string(), z.number()]).describe("Pin number, e.g. '1', '2', 'A1'"),
  type: z
    .enum([
      "input",
      "output",
      "bidirectional",
      "tri_state",
      "passive",
      "free",
      "unspecified",
      "power_in",
      "power_out",
      "open_collector",
      "open_emitter",
      "no_connect",
    ])
    .describe("Electrical pin type"),
  at: z
    .object({
      x: z.number().describe("X position in mm"),
      y: z.number().describe("Y position in mm"),
      angle: z
        .number()
        .describe(
          "Direction the pin wire extends FROM the symbol body: 0=right, 90=up, 180=left, 270=down",
        ),
    })
    .describe("Pin endpoint position (where the wire connects)"),
  length: z.number().optional().describe("Pin length in mm (default 2.54)"),
  shape: z
    .enum([
      "line",
      "inverted",
      "clock",
      "inverted_clock",
      "input_low",
      "clock_low",
      "output_low",
      "falling_edge_clock",
      "non_logic",
    ])
    .optional()
    .describe("Pin graphic shape (default: line)"),
});

const RectSchema = z.object({
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
  width: z.number().optional().describe("Stroke width in mm (default 0.254)"),
  fill: z
    .enum(["none", "outline", "background"])
    .optional()
    .describe("Fill type (default: background)"),
});

const PolylineSchema = z.object({
  points: z.array(z.object({ x: z.number(), y: z.number() })).describe("List of XY points in mm"),
  width: z.number().optional().describe("Stroke width in mm (default 0.254)"),
  fill: z.enum(["none", "outline", "background"]).optional(),
});

export function registerSymbolCreatorTools(server: McpServer, callKicadScript: Function) {
  // ── create_symbol ────────────────────────────────────────────────────── //
  server.tool(
    "create_symbol",
    "Create a new schematic symbol in a .kicad_sym library file (created if missing). " +
      "After creation, use register_symbol_library so KiCAD finds it. " +
      "Pin positions are where the wire connects; the symbol body is drawn between them.\n\n" +
      "Coordinate tips:\n" +
      "- Body rectangle typically spans ±2.54 to ±5.08 mm\n" +
      "- Pins on left side: at.x = body_left - length, angle=0 (wire goes right)\n" +
      "- Pins on right side: at.x = body_right + length, angle=180 (wire goes left)\n" +
      "- Pins on top: at.y = body_top + length, angle=270 (wire goes down)\n" +
      "- Pins on bottom: at.y = body_bottom - length, angle=90 (wire goes up)\n" +
      "- Standard pin length: 2.54 mm, standard grid: 2.54 mm",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file (created if missing)"),
      name: z.string().describe("Symbol name, e.g. 'TMC2209', 'MyOpAmp'"),
      referencePrefix: z
        .string()
        .optional()
        .describe(
          "Schematic reference prefix: 'U' (IC), 'R' (resistor), 'J' (connector), etc. Default: 'U'",
        ),
      description: z.string().optional().describe("Human-readable description"),
      keywords: z.string().optional().describe("Space-separated search keywords"),
      datasheet: z.string().optional().describe("Datasheet URL or '~'"),
      footprint: z
        .string()
        .optional()
        .describe("Default footprint, e.g. 'Package_SO:SOIC-8_3.9x4.9mm_P1.27mm'"),
      inBom: z.boolean().optional().describe("Include in BOM (default true)"),
      onBoard: z.boolean().optional().describe("Include in netlist for PCB (default true)"),
      pins: z
        .array(PinSchema)
        .optional()
        .describe("List of pins (can be empty for graphical-only symbols)"),
      rectangles: z
        .array(RectSchema)
        .optional()
        .describe("Body rectangle(s). Typically one rectangle defining the IC body."),
      polylines: z
        .array(PolylineSchema)
        .optional()
        .describe("Polyline graphics for custom body shapes (op-amp triangles, etc.)"),
      overwrite: z
        .boolean()
        .optional()
        .describe("Replace existing symbol with same name (default false)"),
    },
    async (args: {
      libraryPath: string;
      name: string;
      referencePrefix?: string;
      description?: string;
      keywords?: string;
      datasheet?: string;
      footprint?: string;
      inBom?: boolean;
      onBoard?: boolean;
      pins?: z.infer<typeof PinSchema>[];
      rectangles?: z.infer<typeof RectSchema>[];
      polylines?: z.infer<typeof PolylineSchema>[];
      overwrite?: boolean;
    }) => {
      const result = await callKicadScript("create_symbol", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── delete_symbol ────────────────────────────────────────────────────── //
  server.tool(
    "delete_symbol",
    "Remove a symbol from a .kicad_sym library file.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
      name: z.string().describe("Symbol name to delete"),
    },
    async (args: { libraryPath: string; name: string }) => {
      const result = await callKicadScript("delete_symbol", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── list_symbols_in_library ──────────────────────────────────────────── //
  server.tool(
    "list_symbols_in_library",
    "List all symbol names in a .kicad_sym library file.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
    },
    async (args: { libraryPath: string }) => {
      const result = await callKicadScript("list_symbols_in_library", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── register_symbol_library ──────────────────────────────────────────── //
  server.tool(
    "register_symbol_library",
    "Register a .kicad_sym library in KiCAD's sym-lib-table so symbols can be used in schematics. " +
      "Run this after create_symbol when KiCAD shows 'library not found'.",
    {
      libraryPath: z.string().describe("Full path to the .kicad_sym file"),
      libraryName: z
        .string()
        .optional()
        .describe("Nickname (default: file name without extension)"),
      description: z.string().optional(),
      scope: z
        .enum(["project", "global"])
        .optional()
        .describe("project = writes sym-lib-table next to .kicad_pro; global = user config"),
      projectPath: z
        .string()
        .optional()
        .describe("Path to .kicad_pro or its directory (for scope=project)"),
    },
    async (args: {
      libraryPath: string;
      libraryName?: string;
      description?: string;
      scope?: "project" | "global";
      projectPath?: string;
    }) => {
      const result = await callKicadScript("register_symbol_library", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── add_symbol_property ───────────────────────────────────────────────── //
  server.tool(
    "add_symbol_property",
    "Add or update a custom property (Manufacturer, MPN, LCSC, etc.) on a symbol in a .kicad_sym " +
      "library file. The property is written on the symbol itself rather than on one of its " +
      "units, so multi-unit symbols are handled correctly.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
      symbolName: z.string().describe("Symbol name"),
      propertyName: z.string().describe("Property name (e.g. Manufacturer, MPN)"),
      propertyValue: z.string().describe("Property value"),
      position: z
        .object({ x: z.number(), y: z.number() })
        .optional()
        .describe(
          "Position {x, y} in mm. Omit to leave an existing property where it is (new: 0, 0)",
        ),
      hide: z
        .boolean()
        .optional()
        .describe("Hide the property. Omit to keep an existing property's visibility (new: false)"),
    },
    async (args: any) => {
      const r = await callKicadScript("add_symbol_property", args);
      if (r.success === false)
        return { content: [{ type: "text", text: `Failed: ${r.message || "Unknown error"}` }] };
      return { content: [{ type: "text", text: r.message }] };
    },
  );

  // ── import_symbol ─────────────────────────────────────────────────────── //
  server.tool(
    "import_symbol",
    "Copy a symbol from one .kicad_sym library into another, with optional rename and overwrite. " +
      "The target library is created if missing. A derived symbol (one using (extends ...)) needs " +
      "its parent imported into the target first.",
    {
      sourceLibraryPath: z.string().describe("Path to the source .kicad_sym file"),
      symbolName: z.string().describe("Symbol to import"),
      targetLibraryPath: z.string().describe("Path to the target .kicad_sym (created if missing)"),
      newName: z.string().optional().describe("Rename the symbol on import"),
      overwrite: z
        .boolean()
        .optional()
        .default(false)
        .describe("Overwrite if the symbol already exists in the target"),
    },
    async (args: any) => {
      const result = await callKicadScript("import_symbol", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── export_symbol ─────────────────────────────────────────────────────── //
  server.tool(
    "export_symbol",
    "Extract a single symbol from a .kicad_sym library into a standalone .kicad_sym file.",
    {
      libraryPath: z.string().describe("Path to the source .kicad_sym file"),
      symbolName: z.string().describe("Symbol to export"),
      outputPath: z.string().describe("Path for the output .kicad_sym file"),
    },
    async (args: any) => {
      const result = await callKicadScript("export_symbol", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );

  // ── rename_symbol ─────────────────────────────────────────────────────── //
  server.tool(
    "rename_symbol",
    "Rename a symbol in a .kicad_sym library, including its sub-symbol shards (name_0_1, ...) and " +
      "any (extends ...) references from derived symbols in the same library. Note: schematics that " +
      "already place the old lib_id are NOT updated — use replace_instance_lib_ids for that.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
      oldName: z.string().describe("Current symbol name"),
      newName: z.string().describe("New symbol name"),
    },
    async (args: any) => {
      const result = await callKicadScript("rename_symbol", args);
      return {
        content: [{ type: "text", text: JSON.stringify(result, null, 2) }],
      };
    },
  );
}

File: KiCAD-MCP-Server/python/commands/symbol_creator.py
"""
Symbol Creator for KiCAD MCP Server

Creates and edits .kicad_sym symbol library files using raw S-Expression text generation.
No sexpdata – pure f-string assembly to guarantee format correctness.

KiCAD 9 .kicad_sym format:
  - Library file starts with (kicad_symbol_lib (version 20241209) ...)
  - Each symbol has a parent block with properties + two sub-symbols:
      SymbolName_0_1  → body graphics (rectangle, polyline, circle, arc)
      SymbolName_1_1  → pins
  - All coordinates in mm, 2.54mm grid typical for schematic symbols
"""

import logging
import os
import re
from pathlib import Path
from typing import Any, Dict, List, Optional

from utils.sexpr_format import escape_sexpr_string

logger = logging.getLogger("kicad_interface")

KICAD9_SYMBOL_LIB_VERSION = "20241209"


def _invalidate_symbol_caches() -> None:
    """Drop the module-level symbol resolution/extraction caches.

    create_symbol / delete_symbol rewrite .kicad_sym files and
    register_symbol_library rewrites the sym-lib-table; the caches in
    dynamic_symbol_loader must not outlive those edits. The mtime guards over
    there cover direct file edits already — this explicit clear additionally
    covers re-pointing an existing library name at a different path, which a
    per-file stat cannot see. Lazy import to avoid any import-order coupling;
    a failed clear must never break the write that just succeeded.
    """
    try:
        from commands.dynamic_symbol_loader import DynamicSymbolLoader

        DynamicSymbolLoader.clear_library_caches()
    except Exception:  # pragma: no cover - defensive
        logger.warning("Could not clear symbol library caches", exc_info=True)


# Pin electrical types
PIN_TYPES = {
    "input",
    "output",
    "bidirectional",
    "tri_state",
    "passive",
    "free",
    "unspecified",
    "power_in",
    "power_out",
    "open_collector",
    "open_emitter",
    "no_connect",
}

# Pin graphic shapes
PIN_SHAPES = {
    "line",
    "inverted",
    "clock",
    "inverted_clock",
    "input_low",
    "clock_low",
    "output_low",
    "falling_edge_clock",
    "non_logic",
}


def _fmt(v: float) -> str:
    return f"{v:g}"


def _esc(s: str) -> str:
    """Escape a value for an S-expression double-quoted token (#336)."""
    return escape_sexpr_string(s)


class SymbolCreator:
    """Creates and edits KiCAD .kicad_sym symbol library files."""

    # ------------------------------------------------------------------ #
    #  create_symbol                                                       #
    # ------------------------------------------------------------------ #

    def create_symbol(
        self,
        library_path: str,
        name: str,
        reference_prefix: str = "U",
        description: str = "",
        keywords: str = "",
        datasheet: str = "~",
        footprint: str = "",
        in_bom: bool = True,
        on_board: bool = True,
        pins: Optional[List[Dict[str, Any]]] = None,
        rectangles: Optional[List[Dict[str, Any]]] = None,
        polylines: Optional[List[Dict[str, Any]]] = None,
        overwrite: bool = False,
    ) -> Dict[str, Any]:
        """
        Add a new symbol to a .kicad_sym library (creates the file if missing).

        Parameters
        ----------
        library_path : str
            Path to the .kicad_sym file (created if missing).
        name : str
            Symbol name, e.g. "TMC2209", "MyOpAmp".
        reference_prefix : str
            Schematic reference prefix, e.g. "U", "R", "J".
        description : str
            Human-readable description.
        keywords : str
            Space-separated keyword string for search.
        datasheet : str
            Datasheet URL or "~".
        footprint : str
            Default footprint, e.g. "Package_SO:SOIC-8".
        in_bom : bool
            Include in BOM (default True).
        on_board : bool
            Include in netlist for PCB (default True).
        pins : list of dicts
            Each pin dict:
              name     (str)   – pin name, e.g. "VCC", "GND", "~" for unnamed
              number   (str)   – pin number, e.g. "1", "A1"
              type     (str)   – electrical type: input|output|bidirectional|
                                  passive|power_in|power_out|tri_state|
                                  open_collector|open_emitter|free|unspecified
              at       (dict)  – {"x": float, "y": float, "angle": float}
                                  angle: 0=right, 90=up, 180=left, 270=down
              length   (float) – pin length in mm (default 2.54)
              shape    (str)   – graphic shape: line|inverted|clock|...
                                  (default "line")
        rectangles : list of dicts or None
            Body rectangles: {"x1","y1","x2","y2", "width"(opt), "fill"(opt)}
            fill: "none"|"outline"|"background" (default "background")
        polylines : list of dicts or None
            {"points": [{"x":float,"y":float},...], "width"(opt), "fill"(opt)}
        overwrite : bool
            Replace existing symbol with same name (default False).

        Returns
        -------
        dict with "success", "library_path", "symbol_name", "pin_count"
        """
        lib_path = Path(library_path)
        if lib_path.suffix.lower() != ".kicad_sym":
            lib_path = lib_path.with_suffix(".kicad_sym")

        lib_path.parent.mkdir(parents=True, exist_ok=True)

        # Load or create library
        if lib_path.exists():
            lib_content = lib_path.read_text(encoding="utf-8")
        else:
            lib_content = (
                f"(kicad_symbol_lib\n"
                f"  (version {KICAD9_SYMBOL_LIB_VERSION})\n"
                f'  (generator "kicad-mcp")\n'
                f'  (generator_version "9.0")\n'
                f")\n"
            )

        # Check for duplicate
        if f'(symbol "{name}"' in lib_content:
            if not overwrite:
                return {
                    "success": False,
                    "error": f'Symbol "{name}" already exists in {lib_path}. Use overwrite=true.',
                    "library_path": str(lib_path),
                }
            lib_content = self._remove_symbol(lib_content, name)

        pins = pins or []
        rectangles = rectangles or []
        polylines = polylines or []

        symbol_block = self._build_symbol_block(
            name=name,
            reference_prefix=reference_prefix,
            description=description,
            keywords=keywords,
            datasheet=datasheet,
            footprint=footprint,
            in_bom=in_bom,
            on_board=on_board,
            pins=pins,
            rectangles=rectangles,
            polylines=polylines,
        )

        # Insert before closing paren of library
        lib_content = lib_content.rstrip()
        if lib_content.endswith(")"):
            lib_content = lib_content[:-1].rstrip() + "\n" + symbol_block + "\n)\n"
        else:
            lib_content += "\n" + symbol_block + "\n)\n"

        lib_path.write_text(lib_content, encoding="utf-8")
        logger.info(f"Created symbol '{name}' in {lib_path} ({len(pins)} pins)")
        _invalidate_symbol_caches()

        return {
            "success": True,
            "library_path": str(lib_path),
            "symbol_name": name,
            "pin_count": len(pins),
        }

    # ------------------------------------------------------------------ #
    #  delete_symbol                                                       #
    # ------------------------------------------------------------------ #

    def delete_symbol(self, library_path: str, name: str) -> Dict[str, Any]:
        """Remove a symbol from a .kicad_sym library."""
        lib_path = Path(library_path)
        if not lib_path.exists():
            return {"success": False, "error": f"Library not found: {library_path}"}

        content = lib_path.read_text(encoding="utf-8")
        if f'(symbol "{name}"' not in content:
            return {"success": False, "error": f'Symbol "{name}" not found in {library_path}'}

        new_content = self._remove_symbol(content, name)
        lib_path.write_text(new_content, encoding="utf-8")
        _invalidate_symbol_caches()
        return {"success": True, "library_path": str(lib_path), "deleted": name}

    # ------------------------------------------------------------------ #
    #  list_symbols  (in a single library file)                           #
    # ------------------------------------------------------------------ #

    def list_symbols(self, library_path: str) -> Dict[str, Any]:
        """List all symbols in a .kicad_sym file."""
        lib_path = Path(library_path)
        if not lib_path.exists():
            return {"success": False, "error": f"Library not found: {library_path}"}

        content = lib_path.read_text(encoding="utf-8")
        # Only top-level symbols (not sub-symbols like _0_1 or _1_1)
        names = re.findall(r'^\s*\(symbol "([^"_][^"]*)"', content, re.MULTILINE)
        # Filter out sub-symbols (contain _N_N suffix)
        symbols = [n for n in names if not re.search(r"_\d+_\d+$", n)]
        return {
            "success": True,
            "library_path": str(lib_path),
            "symbol_count": len(symbols),
            "symbols": symbols,
        }

    # ------------------------------------------------------------------ #
    #  register_symbol_library                                             #
    # ------------------------------------------------------------------ #

    def register_symbol_library(
        self,
        library_path: str,
        library_name: Optional[str] = None,
        description: str = "",
        scope: str = "project",
        project_path: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Register a .kicad_sym library in KiCAD's sym-lib-table.

        Parameters
        ----------
        library_path : str  – path to the .kicad_sym file
        library_name : str  – nickname (default: file stem)
        scope : str         – "project" or "global"
        project_path : str  – .kicad_pro or directory (for scope=project)
        """
        sym_path = Path(library_path)
        name = library_name or sym_path.stem
        uri = str(sym_path).replace("\\", "/")

        if scope == "project":
            if project_path:
                proj = Path(project_path)
                table_dir = proj if proj.is_dir() else proj.parent
            else:
                table_dir = sym_path.parent
            table_path = table_dir / "sym-lib-table"
        else:
            cfg_dirs = [
                Path(os.environ.get("APPDATA", "")) / "kicad" / "9.0",
                Path.home() / ".config" / "kicad" / "9.0",
            ]
            table_path = None
            for d in cfg_dirs:
                candidate = d / "sym-lib-table"
                if candidate.exists():
                    table_path = candidate
                    break
            if table_path is None:
                for d in cfg_dirs:
                    try:
                        d.mkdir(parents=True, exist_ok=True)
                        table_path = d / "sym-lib-table"
                        break
                    except OSError:
                        continue
            if table_path is None:
                return {"success": False, "error": "Could not find/create global sym-lib-table"}

        if table_path.exists():
            content = table_path.read_text(encoding="utf-8")
        else:
            content = "(sym_lib_table\n  (version 7)\n)\n"

        if f'(name "{name}")' in content or uri in content:
            return {
                "success": True,
                "already_registered": True,
                "table_path": str(table_path),
                "library_name": name,
            }

        new_entry = (
            f'  (lib (name "{name}")'
            f'(type "KiCad")'
            f'(uri "{uri}")'
            f'(options "")'
            f'(descr "{_esc(description)}"))'
        )
        content = content.rstrip()
        if content.endswith(")"):
            content = content[:-1].rstrip() + "\n" + new_entry + "\n)\n"
        else:
            content += "\n" + new_entry + "\n)\n"

        table_path.write_text(content, encoding="utf-8")
        logger.info(f"Registered symbol library '{name}' in {table_path}")
        _invalidate_symbol_caches()

        return {
            "success": True,
            "already_registered": False,
            "table_path": str(table_path),
            "library_name": name,
            "uri": uri,
        }

    # ------------------------------------------------------------------ #
    #  Internal helpers                                                    #
    # ------------------------------------------------------------------ #

    def _build_symbol_block(
        self,
        name: str,
        reference_prefix: str,
        description: str,
        keywords: str,
        datasheet: str,
        footprint: str,
        in_bom: bool,
        on_board: bool,
        pins: List[Dict[str, Any]],
        rectangles: List[Dict[str, Any]],
        polylines: List[Dict[str, Any]],
    ) -> str:
        lines: List[str] = []
        bom_str = "yes" if in_bom else "no"
        board_str = "yes" if on_board else "no"

        lines.append(f'  (symbol "{name}"')
        lines.append(f"    (exclude_from_sim no)")
        lines.append(f"    (in_bom {bom_str})")
        lines.append(f"    (on_board {board_str})")

        # Properties
        lines.extend(_property_block("Reference", reference_prefix, 2.54, 0, visible=True))
        lines.extend(_property_block("Value", name, 0, -2.54, visible=True))
        lines.extend(_property_block("Footprint", footprint, 0, -5.08, visible=False))
        lines.extend(_property_block("Datasheet", datasheet or "~", 0, -7.62, visible=False))
        lines.extend(_property_block("Description", description, 0, -10.16, visible=False))
        if keywords:
            lines.extend(_property_block("ki_keywords", keywords, 0, 0, visible=False))

        # Sub-symbol _0_1: body graphics
        lines.append(f'    (symbol "{name}_0_1"')
        for rect in rectangles:
            lines.extend(_rect_sym_lines(rect))
        for pl in polylines:
            lines.extend(_polyline_lines(pl))
        lines.append(f"    )")

        # Sub-symbol _1_1: pins
        lines.append(f'    (symbol "{name}_1_1"')
        for pin in pins:
            lines.extend(_pin_lines(pin))
        lines.append(f"    )")

        lines.append(f"  )")
        return "\n".join(lines)

    def _remove_symbol(self, content: str, name: str) -> str:
        """Remove a complete symbol block from library content."""
        lines = content.split("\n")
        result = []
        skip = False
        depth = 0

        for line in lines:
            stripped = line.strip()
            if not skip:
                if re.match(rf'^\s*\(symbol "{re.escape(name)}"', line) and not re.search(
                    r'_\d+_\d+"', line
                ):
                    skip = True
                    depth = stripped.count("(") - stripped.count(")")
                    continue
                result.append(line)
            else:
                depth += stripped.count("(") - stripped.count(")")
                if depth <= 0:
                    skip = False

        return "\n".join(result)


# ------------------------------------------------------------------ #
#  S-Expression helper functions                                       #
# ------------------------------------------------------------------ #


def _property_block(key: str, value: str, x: float, y: float, visible: bool = True) -> List[str]:
    hide = "" if visible else "\n      (hide yes)"
    return [
        f'    (property "{_esc(key)}" "{_esc(value)}"',
        f"      (at {_fmt(x)} {_fmt(y)} 0)",
        f"      (effects",
        f"        (font (size 1.27 1.27))",
        f"      ){hide}",
        f"    )",
    ]


def _rect_sym_lines(rect: Dict[str, Any]) -> List[str]:
    x1 = _fmt(rect.get("x1", -2.54))
    y1 = _fmt(rect.get("y1", -2.54))
    x2 = _fmt(rect.get("x2", 2.54))
    y2 = _fmt(rect.get("y2", 2.54))
    w = _fmt(rect.get("width", 0.254))
    fill = rect.get("fill", "background")
    return [
        f"      (rectangle",
        f"        (start {x1} {y1})",
        f"        (end {x2} {y2})",
        f"        (stroke (width {w}) (type default))",
        f"        (fill (type {fill}))",
        f"      )",
    ]


def _polyline_lines(pl: Dict[str, Any]) -> List[str]:
    pts = pl.get("points", [])
    w = _fmt(pl.get("width", 0.254))
    fill = pl.get("fill", "none")
    lines = [
        f"      (polyline",
        f"        (pts",
    ]
    for pt in pts:
        lines.append(f'          (xy {_fmt(pt["x"])} {_fmt(pt["y"])})')
    lines += [
        f"        )",
        f"        (stroke (width {w}) (type default))",
        f"        (fill (type {fill}))",
        f"      )",
    ]
    return lines


def _pin_lines(pin: Dict[str, Any]) -> List[str]:
    ptype = pin.get("type", "passive").lower()
    shape = pin.get("shape", "line").lower()
    at = pin.get("at", {"x": 0, "y": 0, "angle": 0})
    x = _fmt(at.get("x", 0))
    y = _fmt(at.get("y", 0))
    angle = _fmt(at.get("angle", 0))
    length = _fmt(pin.get("length", 2.54))
    pin_name = pin.get("name", "~")
    pin_number = str(pin.get("number", "1"))

    return [
        f"      (pin {ptype} {shape}",
        f"        (at {x} {y} {angle})",
        f"        (length {length})",
        f'        (name "{_esc(pin_name)}"',
        f"          (effects (font (size 1.27 1.27)))",
        f"        )",
        f'        (number "{_esc(pin_number)}"',
        f"          (effects (font (size 1.27 1.27)))",
        f"        )",
        f"      )",
    ]
